feat(search): Add unaccent support for accent-insensitive search.

mabdullahabaid opened the pull request:
## ✨ Add accent-insensitive search functionality

### 🎯 Overview
Implements accent-insensitive search across all searchable fields in Twenty CRM.  
Users can now search for "jose" to find "José", "muller" to find "Müller", "cafe" to find "café", etc.

### 🔍 Problem
Twenty's search functionality was accent-sensitive, requiring users to type exact accented characters to find records.  
This created a poor user experience, especially for international names and content.

### 💡 Solution
Added PostgreSQL `unaccent` extension with a custom immutable wrapper function to enable accent-insensitive full-text search across all searchable field types.

### 📋 Changes Made
**Modified Files:**  
- `packages/twenty-server/scripts/setup-db.ts`  
- `packages/twenty-server/src/engine/api/graphql/graphql-query-runner/utils/compute-where-condition-parts.ts`  
- `packages/twenty-server/src/engine/workspace-manager/workspace-sync-metadata/utils/get-ts-vector-column-expression.util.ts`

### 🗄️ Database Setup (`setup-db.ts`)
```sql
-- Added unaccent extension
CREATE EXTENSION IF NOT EXISTS "unaccent";

-- Created immutable wrapper function
CREATE OR REPLACE FUNCTION unaccent_immutable(text) RETURNS text AS $$
  SELECT public.unaccent($1)
$$ LANGUAGE sql IMMUTABLE;
```

### 🔍 Search Vector Generation (`get-ts-vector-column-expression.util.ts`)
Applied `public.unaccent_immutable()` to all searchable field types:  
- TEXT fields (job titles, names, etc.)  
- FULL_NAME fields (first/last names)  
- EMAILS fields (both email address and domain)  
- ADDRESS fields  
- LINKS fields  
- RICH_TEXT and RICH_TEXT_V2 fields

### 🔎 Query Processing (`compute-where-condition-parts.ts`)
Enhanced search queries to use `public.unaccent_immutable()` for both:  
- Full-text search (`@@` operator with `to_tsquery`)  
- Pattern matching (`ILIKE` operator)

### 🧠 Technical Rationale: Why the Wrapper Function?
**The Challenge:**  
PostgreSQL's built-in `unaccent()` is marked as **STABLE**, but `GENERATED ALWAYS AS` expressions (used for search vector columns) require **IMMUTABLE** functions.

**The Solution:**  
Created an IMMUTABLE wrapper function that calls the underlying `unaccent()` function:  
- ✅ Satisfies PostgreSQL's immutability requirements for generated columns  
- ✅ Maintains the exact same functionality as the original `unaccent()`  
- ✅ Uses fully qualified `public.unaccent_immutable()` to ensure function resolution from workspace schemas

**Alternative Approaches Considered:**  
- ❌ Modifying `search_path`: would affect workspace isolation  
- ❌ Computing unaccent at query time: would hurt performance  
- ❌ Using triggers: would complicate data consistency

### 🎯 Impact
For **Person** records, accent-insensitive search now works on:  
- Name (first/last name): `"jose garcia"` finds `"José García"`  
- Email: `"jose@cafe.com"` finds `"josé@café.com"`  
- Job Title: `"manager"` finds `"Managér"` or `"Gerente de Café"`

Applies to all searchable standard objects:  
- Companies, People, Opportunities, Notes, Tasks, etc.  
- Any custom fields of searchable types (TEXT, EMAILS, etc.)

### ✅ Testing
- Database reset completes successfully  
- Workspace seeding works without errors  
- Search vectors generate with unaccent functionality  
- All searchable field types properly handle accented characters

FelixMalfait commented on packages/twenty-server/scripts/setup-db.ts:
```diff
@@ -21,6 +21,18 @@ rawDataSource
+
+    await performQuery(
+      `CREATE OR REPLACE FUNCTION unaccent_immutable(text) RETURNS text AS $$
+        SELECT public.unaccent($1)
```
What would happen if unaccent had not been installed? Ex: some postgres providers might not allow installing it. Would this degrade gracefully?

FelixMalfait replied:
Also is this always created in the public schema? I have the intuition that yes because it works but I can't tell why

FelixMalfait replied:
Yes I'm not sure if we should do that IF statement at the Postgres level or at the server-environment level...

But now I check and it seems to be very widely supported:
<img width="598" height="197" alt="Screenshot 2025-09-13 at 18 50 15" src="https://github.com/user-attachments/assets/4bbe2ebf-e5e7-460f-8baa-0b5e008e2b97" />

So let's go ahead and merge the way you did. If someone complains we'll add an environment variable

mabdullahabaid replied:
Sounds good.

FelixMalfait commented on packages/twenty-server/test/integration/search/unaccent-search.integration-spec.ts:
```diff
@@ -0,0 +1,203 @@
+      `);
+
+      // Insert test data
+      await global.testDataSource.query(`
```
Theses tests are too low-level for integration tests. You could say it doesn't hurt but I'm worried merging this will set a bad precedent for what inegration tests are in the code base. 

As you see in the test I sent you, it usually call `makeGraphqlAPIRequest` so it's at a much higher level. You don't interact with the database but instead test and go through the graphql layer, etc

Did you try it and hit a roadblock?

FelixMalfait replied:
No problem at all. Your tests were interesting! Thanks for the great work!

FelixMalfait reviewed (commented):
Sorry one last comment on integration tests then we can merge!

FelixMalfait commented:
Actually I tested it and it seems to account for one case but not the other. Input should also me modified to remove accents.

<img width="1183" height="174" alt="Screenshot 2025-09-14 at 08 17 35" src="https://github.com/user-attachments/assets/ec8f1d06-ee56-4dea-a079-795dd04fb1ab" />

<img width="1182" height="175" alt="Screenshot 2025-09-14 at 08 17 47" src="https://github.com/user-attachments/assets/355e3827-e8b0-4778-81d1-3f2693c2ea85" />

mabdullahabaid commented:
I went back into the codebase, figured I need to update the integration test named `search-resolver.integration-spec.ts` instead of introducing a new one, so I removed the specific integration test I introduced for unaccent with direct database connection.

As discussed on Discord, I shall fix the integration test during the early part of this week given the time constraints However, here are the screenshots confirming that input has been fixed as requested.

These records were inserted using `psql` with `docker exec`, so the `Created by` field was assigned `System` instead of `Tim Apple`.

When testing, you may also find records that do not contain the search term. That is because the search term is a part of the `note` field - it's the existing functionality, not something I changed.

<img width="1919" height="464" alt="Screenshot 2025-09-14 181225" src="https://github.com/user-attachments/assets/3410d0df-51cd-49e3-a2c4-cd14050a6a0f" />

<img width="1919" height="458" alt="Screenshot 2025-09-14 181303" src="https://github.com/user-attachments/assets/5aa9ead0-c43b-4377-bfd9-15a76f6865b5" />

<img width="1918" height="458" alt="Screenshot 2025-09-14 181346" src="https://github.com/user-attachments/assets/bad10ebe-c58a-42df-9184-65c1573168f9" />

<img width="1919" height="462" alt="Screenshot 2025-09-14 181409" src="https://github.com/user-attachments/assets/59fcd32b-48a1-48fe-9752-0037d2ef284c" />

<img width="1919" height="426" alt="Screenshot 2025-09-14 181459" src="https://github.com/user-attachments/assets/512ca432-500f-4d31-bcfe-889fece2da49" />

<img width="1918" height="422" alt="Screenshot 2025-09-14 181621" src="https://github.com/user-attachments/assets/9aa911db-b133-49b3-ad29-be123eaa1684" />

<img width="1919" height="422" alt="Screenshot 2025-09-14 181716" src="https://github.com/user-attachments/assets/a35c9d1d-2b85-4494-8311-a95ccdd752d6" />

<img width="1917" height="460" alt="Screenshot 2025-09-14 181748" src="https://github.com/user-attachments/assets/5dae8d6c-1e8b-43fd-bc00-d1417bc27bc9" />

<img width="1918" height="457" alt="Screenshot 2025-09-14 181818" src="https://github.com/user-attachments/assets/9bdd7df2-28bd-4d36-9888-d2d00dabe90c" />

FelixMalfait commented on packages/twenty-server/src/engine/core-modules/search/utils/unaccent-text.ts:
```diff
@@ -0,0 +1,34 @@
+  let result = text;
+
+  for (const [accented, unaccented] of Object.entries(specialChars)) {
+    result = result.replace(new RegExp(accented, 'g'), unaccented);
```
Ignore this comment since implementation should rather rely on Postgres' unaccent

FelixMalfait commented on packages/twenty-server/src/engine/core-modules/search/utils/unaccent-text.ts:
```diff
@@ -0,0 +1,34 @@
+export const unaccentText = (text: string): string => {
```
If you don't manage to do it at the Postgres-query level, then instead of building you own implementation, I would look at the most common unaccent.rules postgres implementation and hardcode the same rule instead. But that's really not as good!

E.g.

```
const POSTGRESQL_UNACCENT_RULES: Record<string, string> = {
  // Latin-1 Supplement
  'À': 'A', 'Á': 'A', 'Â': 'A', 'Ã': 'A', 'Ä': 'A', 'Å': 'A', 'Æ': 'AE',
  'Ç': 'C', 'È': 'E', 'É': 'E', 'Ê': 'E', 'Ë': 'E', 'Ì': 'I', 'Í': 'I',
  'Î': 'I', 'Ï': 'I', 'Ð': 'D', 'Ñ': 'N', 'Ò': 'O', 'Ó': 'O', 'Ô': 'O',
  'Õ': 'O', 'Ö': 'O', 'Ø': 'O', 'Ù': 'U', 'Ú': 'U', 'Û': 'U', 'Ü': 'U',
  'Ý': 'Y', 'Þ': 'TH', 'ß': 'ss', 'à': 'a', 'á': 'a', 'â': 'a', 'ã': 'a',
  'ä': 'a', 'å': 'a', 'æ': 'ae', 'ç': 'c', 'è': 'e', 'é': 'e', 'ê': 'e',
  'ë': 'e', 'ì': 'i', 'í': 'i', 'î': 'i', 'ï': 'i', 'ð': 'd', 'ñ': 'n',
  'ò': 'o', 'ó': 'o', 'ô': 'o', 'õ': 'o', 'ö': 'o', 'ø': 'o', 'ù': 'u',
  'ú': 'u', 'û': 'u', 'ü': 'u', 'ý': 'y', 'þ': 'th', 'ÿ': 'y',

  // Latin Extended-A
  'Ā': 'A', 'ā': 'a', 'Ă': 'A', 'ă': 'a', 'Ą': 'A', 'ą': 'a', 'Ć': 'C',
  'ć': 'c', 'Ĉ': 'C', 'ĉ': 'c', 'Ċ': 'C', 'ċ': 'c', 'Č': 'C', 'č': 'c',
  'Ď': 'D', 'ď': 'd', 'Đ': 'D', 'đ': 'd', 'Ē': 'E', 'ē': 'e', 'Ĕ': 'E',
  'ĕ': 'e', 'Ė': 'E', 'ė': 'e', 'Ę': 'E', 'ę': 'e', 'Ě': 'E', 'ě': 'e',
  'Ĝ': 'G', 'ĝ': 'g', 'Ğ': 'G', 'ğ': 'g', 'Ġ': 'G', 'ġ': 'g', 'Ģ': 'G',
  'ģ': 'g', 'Ĥ': 'H', 'ĥ': 'h', 'Ħ': 'H', 'ħ': 'h', 'Ĩ': 'I', 'ĩ': 'i',
  'Ī': 'I', 'ī': 'i', 'Ĭ': 'I', 'ĭ': 'i', 'Į': 'I', 'į': 'i', 'İ': 'I',
  'ı': 'i', 'Ĳ': 'IJ', 'ĳ': 'ij', 'Ĵ': 'J', 'ĵ': 'j', 'Ķ': 'K', 'ķ': 'k',
  'ĸ': 'k', 'Ĺ': 'L', 'ĺ': 'l', 'Ļ': 'L', 'ļ': 'l', 'Ľ': 'L', 'ľ': 'l',
  'Ŀ': 'L', 'ŀ': 'l', 'Ł': 'L', 'ł': 'l', 'Ń': 'N', 'ń': 'n', 'Ņ': 'N',
  'ņ': 'n', 'Ň': 'N', 'ň': 'n', 'ŉ': 'n', 'Ŋ': 'N', 'ŋ': 'n', 'Ō': 'O',
  'ō': 'o', 'Ŏ': 'O', 'ŏ': 'o', 'Ő': 'O', 'ő': 'o', 'Œ': 'OE', 'œ': 'oe',
  'Ŕ': 'R', 'ŕ': 'r', 'Ŗ': 'R', 'ŗ': 'r', 'Ř': 'R', 'ř': 'r', 'Ś': 'S',
  'ś': 's', 'Ŝ': 'S', 'ŝ': 's', 'Ş': 'S', 'ş': 's', 'Š': 'S', 'š': 's',
  'Ţ': 'T', 'ţ': 't', 'Ť': 'T', 'ť': 't', 'Ŧ': 'T', 'ŧ': 't', 'Ũ': 'U',
  'ũ': 'u', 'Ū': 'U', 'ū': 'u', 'Ŭ': 'U', 'ŭ': 'u', 'Ů': 'U', 'ů': 'u',
  'Ű': 'U', 'ű': 'u', 'Ų': 'U', 'ų': 'u', 'Ŵ': 'W', 'ŵ': 'w', 'Ŷ': 'Y',
  'ŷ': 'y', 'Ÿ': 'Y', 'Ź': 'Z', 'ź': 'z', 'Ż': 'Z', 'ż': 'z', 'Ž': 'Z',
  'ž': 'z', 'ſ': 's',
};

const UNACCENT_REGEX = new RegExp(`[${Object.keys(POSTGRESQL_UNACCENT_RULES).join('')}]`, 'g');
```

mabdullahabaid replied:
Sure, I will take this up again on Tuesday.

FelixMalfait replied:
Thank you!

mabdullahabaid commented:
Fixed the logic to do unaccent for input at the database-level too. Here are some screenshots to confirm.

Yet to update the integration test, ran out of time today.

<img width="1917" height="944" alt="Screenshot 2025-09-17 051042" src="https://github.com/user-attachments/assets/946089e6-aaa9-4ff5-ab0a-e2313fddbcdd" />

<img width="1918" height="943" alt="Screenshot 2025-09-17 051121" src="https://github.com/user-attachments/assets/0cce4234-08d7-40ba-bc42-19e43848ec69" />

<img width="1919" height="941" alt="Screenshot 2025-09-17 051201" src="https://github.com/user-attachments/assets/ae020bf4-190f-4850-a2f4-e5a063551dd0" />

<img width="1919" height="941" alt="Screenshot 2025-09-17 051227" src="https://github.com/user-attachments/assets/9bfed3a6-e262-4c61-a217-a09981af22f7" />

<img width="1919" height="943" alt="Screenshot 2025-09-17 051313" src="https://github.com/user-attachments/assets/71c1943d-48d8-44fc-ad52-b7e172c5b7b3" />

<img width="1919" height="944" alt="Screenshot 2025-09-17 051350" src="https://github.com/user-attachments/assets/aa426a77-b83c-4092-a8e0-2595296e6b0c" />

<img width="1919" height="942" alt="Screenshot 2025-09-17 051456" src="https://github.com/user-attachments/assets/a00ba935-8e14-4bdc-805f-3732bf3ad226" />

<img width="1916" height="944" alt="Screenshot 2025-09-17 051526" src="https://github.com/user-attachments/assets/8e3728e7-d4ea-46d3-8d56-a8132a4c1e4d" />
